### ISSUE TITLE
Custom Functions are not working with Iterative Calculation

### ISSUE THREAD

**Issue by 4tti** [OfficeDev/office-js#4738]
# Provide required information needed to triage your issue

## Your Environment

* Platform [PC desktop, Mac, iOS, Office on the web]: PC Desktop
* Host [Excel, Word, PowerPoint, etc.]: Excel
* Office version number: 2405 build 17628.20188
* Operating System: Windows
 
## Expected behavior

Users can use iterative calculation with Custom Functions.

## Current behavior

Custom Functions return #CALC! error

## Steps to reproduce

1. In C2 add this formula "=IF(C2 >5;1;C2+1)"
2. Enable iterative calculation: 
![image](https://github.com/user-attachments/assets/b0a22773-fc63-473e-80f2-c7a03704888a)
3. Register script lab's default custom function BASICCUSTOMFUNCTION.SPHEREVOLUME
4. In cell C6 insert "=SCRIPTLAB.BASICCUSTOMFUNCTION.SPHEREVOLUME(C2)"

Result will be #CALC!

## Context

Iterative calculation is powerful feature and alows customer e.g. loop through some data and use it for ETL. Now this is completely blocked.

**Comment by shanshanzheng-dev** [comment]
@4tti Thanks for reporting this issue, I can repro it. We'll investigate and report back if we have a suggestion for you. Thanks.

**Comment by 4tti** [comment]
Thanks a lot @shanshanzheng-dev.
Just want to emphasize, that iterative calculation with UDFs (old type of functions) was heavily used in certain scenarios and with CustomFunctions not working we are blocked to migrate such scenarios.

**Comment by hdwatts** [comment]
This is an issue for us as well, hoping this can be resolved quickly!

**Comment by cjtafoya** [comment]
+1 this is a fundamental blocker -- iterative calculations are required in all basic financial models (interest expense), so this is a MASSIVE problem for any finance-related OfficeJS implementation

**Comment by XuanZhouMSFT** [comment]
Thanks for reporting this issue regarding custom function in iterative calculation scenarios. This is a currently limitation for CF.
Aware of so many complaints, we already put it on our backlog <9225466> to get rid of this limitation. And we will keep you updated.

**Comment by 4tti** [comment]
Thank you for the update @XuanZhouMSFT

**Comment by hdwatts** [comment]
@shanshanzheng-dev @penglongzhaochina Checking in on the status of this issue, do you have an update for us?

**Comment by penglongzhaochina** [comment]
Hi it's a feature level task. Will let you know once we have any progress.

**Comment by evelyn-chrono** [comment]
@penglongzhaochina @qinliuMSFT @shanshanzheng-dev  Any updates here? This is extremely high priority for our users.

**Comment by hdwatts** [comment]
@penglongzhaochina @qinliuMSFT This is a critical issue for us that is causing a negative experience for our end users. Iterative calculations are required in all basic financial models - can you please share an update here?

**Comment by XuanZhouMSFT** [comment]
@evelyn-chrono @hdwatts, thanks all for the waiting. We already started to work on this item. And we will update here once it released.

**Comment by evelyn-chrono** [comment]
@XuanZhouMSFT Any updates here? We are still blocked by this issue.

**Comment by evelyn-chrono** [comment]
@penglongzhaochina @qinliuMSFT  @XuanZhouMSFT checking in again, any updates?

**Comment by hdwatts** [comment]
@trigurl / @XuanZhouMSFT Has this item really had work started on it? Or is it still in the backlog? If work has really been started can we adjust the issue labels to reflect as such?

**Comment by argosy** [comment]
I don't think MSFT truly understands how important it is for custom functions to work in iterative mode.  There is currently NO way to protect your intellectual property in excel, as the password protection in VBA can be broken inside a minute.  The only way to protect proprietary algorithms is to hide them in server-side code (as far as I know).

**Comment by evelyn-chrono** [comment]
@penglongzhaochina @RuizhiSunMS has there been any update on the status here? the last thing we heard was that you started work on this more than 6 months ago.

**Comment by RuizhiSunMS** [comment]
Hi @argosy @evelyn-chrono @hdwatts 
Thx for your attentions and patiences. Our owning feature crew has worked on a new feature for some time to make CF support sync functionalities. Supporting iterative calculation is also in the scope of plan. 
Although we can't share a explicit timeline, we will continue to update if we got any achievement

**Comment by hdwatts** [comment]
@RuizhiSunMS @AmandaHan123 any update to share here?

**Comment by evelyn-chrono** [comment]
@amagupta-collab @RuizhiSunMS @ranshul-microsoft @AmandaHan123 @pravkuma Any updates here?

**Comment by evelyn-chrono** [comment]
@amagupta-collab @RuizhiSunMS @ranshul-microsoft @AmandaHan123 @pravkuma Have there been any updates here?

**Comment by ayush-meghwani-28** [comment]
Hello @4tti, thanks for raising this!

You're right — this is a known limitation. Custom Functions currently return #CALC! when used inside iterative calculation workbooks, and we have this tracked internally as a planned feature enhancement. We expect to address it in a future update, though we're not able to commit to a specific timeline just yet.

We understand this could be a blocker for certain scenarios. Could you share a bit more about your use case? Depending on what you're trying to achieve, there may be a workaround that can help you in the meantime while we work toward full support.

If this is a hard blocker and you'd like to formally signal the need to our product team, we'd also encourage you to submit or upvote a request on the Microsoft Tech Community Ideas Forum — it helps us gauge community interest and prioritize accordingly:

🔗 [Microsoft 365 Developer Platform Ideas](https://techcommunity.microsoft.com/t5/microsoft-365-developer-platform/idb-p/Microsoft365Developer)

Thank you for your patience and for taking the time to report this!

**Comment by 4tti** [comment]
hello @ayush-meghwani-28 , It is not about me but about customers. What exactly do you need to understand?
There are many what-if scenarios and iterative calculation scenarios that are pretty usefull for financial/data analysis scenarios. Just ask any LLM for the use cases and it will throw some of them.

**Comment by ayush-meghwani-28** [comment]
Hello @4tti, thanks for your response.

I understand your concern, and I want to reassure you that this issue is being tracked internally. Work toward supporting these scenarios is moving in the right direction and continuing forward. We just can’t commit to a fix deadline or to when this feature will be available.

What I wanted to share is that this feedback has been noted and is already in the design and implementation plans. If this is a hard-core and super blocking scenario for you, please use the Ideas forum to track it and help it gain traction so PM's can triage it with even higher priority.

This will be handled in upcoming future work. Thanks for understanding!

**Comment by 4tti** [comment]
I think the users are using 'workarounds' replacing the custom functions with something else (values), IF(TRUE, Hardcoded Value, CustomFunction), etc. but it is not optimal at all, because they need to undergo various steps to get it working and doing some of the updates manually.
Anyway, thanks a lot.

**Comment by ayush-meghwani-28** [comment]
Yes, even the performative work for custom functions is underway where all the DIM (delete insert and modified) operations which happens on excel update (insert row) are optimized in certain way.

Hope we will be able to serve better to all the consumers in futures but thanks for sharing the feedback!

**Comment by 4tti** [comment]
Maybe @hdwatts @cjtafoya  can share more details...

**Comment by evelyn-chrono** [comment]
@ayush-meghwani-28 To add more context on the issue, our custom functions pull financial data into Excel models (GET-style functions) and write model outputs back to our platform (PUSH-style functions). 

As an example use case: standard leveraged-finance models are inherently circular. Interest expense is computed on the average debt balance, but the ending balance depends on cash available for paydown, which depends on interest expense. The same applies to revolver/cash-sweep modeling and fund fee/carry waterfalls (fees on NAV, where NAV is net of fees). Analysts resolve these with iterative calculation; it's been standard practice and worked fine with XLL/VBA UDFs.

I've added a post to the Ideas forum so that your team can track this feedback.
https://techcommunity.microsoft.com/idea/microsoft365developerplatform/support-custom-functions-downstream-of-iterative-circular-calculation---currentl/4536519

To confirm the exact issue we're seeing, any custom function consuming a circular model's outputs (e.g. an add-in function that writes results back to an external platform) is by definition downstream of the circular chain, so it always fails in these models. We've verified: a custom function with literal arguments works; the same function with an argument fed from the converged loop returns #CALC!; severing the dependency with IF(TRUE, value, FN(...)) makes it work again. It isn't that enabling iterative calcs breaks all custom functions in the workbook, it's that custom functions dependent on an circular chain resolved with iterative calcs return #CALC!.  

The fix we'd hope for is that custom functions downstream of a circular chain evalaute against the loop's converged values once iteration completes, matching the legacy XLL/VBA UDF behavior. Custom functions do not need to participate in the iteration itself.

@4tti @argosy tagging you both so you can update the ideas thread

**Comment by ayush-meghwani-28** [comment]
hello @evelyn-chrono , thanks for sharing this added context.

We will make sure to use this context while testing out the iterative calculation feature but as said earlier this is already being tracked in our feature backlog and we are committed to deliver this feature soon and hoping to unblock all of these scenarios.

Thanks!

**Comment by evelyn-chrono** [comment]
@ayush-meghwani-28 Is more author feedback needed? I want to make sure this issue isn't closed incorrectly. I think the issue is fairly clear at this point.

**Comment by ayush-meghwani-28** [comment]
Hello @evelyn-chrono Nope, there is no feedback required and this feature as said earlier has already been on track internally and I said that I can't comment on any specific release date, but it is expected to release starting 2027. 
And since there is no issue or bug as such, I was hoping to close this GitHub-issue ticket as right platform to track feature is Ideas forum, and we are tracking this internally as well.
But I will leave up to you if you want to keep this ticket open as your way of tracking the item else, we can close this. Thanks for all the feedback.
